tests/libwait/wait.go

```diff
@@ -39,25 +39,89 @@ import (
 	"kubevirt.io/kubevirt/tests/watcher"
 )
 
-// WaitForVMIStartOrFailed blocks until the specified VirtualMachineInstance reached either Failed or Running states
-func WaitForVMIStartOrFailed(obj runtime.Object, seconds int, wp watcher.WarningsPolicy) *v1.VirtualMachineInstance {
-	ctx, cancel := context.WithCancel(context.Background())
-	defer cancel()
-	return waitForVMIPhase(ctx, []v1.VirtualMachineInstancePhase{v1.Running, v1.Failed}, obj, seconds, wp, true)
+const defaultTimeout = 360
+
+// Option represents an action that enables an option.
+type Option func(waiting Waiting)
+
+// Waiting represents the waiting struct.
+type Waiting struct {
+	ctx         context.Context
+	wp          *watcher.WarningsPolicy
+	timeout     int
+	waitForFail bool
+}
+
+// WaitingFor instantiates a Waiting configuration and return the vmi,
+// building its properties based on the specified With* options.
+func WaitingFor(obj runtime.Object, phases []v1.VirtualMachineInstancePhase, opts ...Option) *v1.VirtualMachineInstance {
+	wp := watcher.WarningsPolicy{FailOnWarnings: true}
+	waiting := Waiting{timeout: defaultTimeout, wp: &wp}
+	for _, f := range opts {
+		f(waiting)
+	}
+
+	if waiting.ctx == nil {
+		ctx, cancel := context.WithCancel(context.Background())
```

Comment:
Consider using the `WithContext` option and not directly.

Well, IMO a context should not be optional or default, it should always be explicit.
But I recognize this "rule" can be soften for tests (at least until I encounter a flake who will change my mind :) ).

Reply:
Initially I was using `WithContext` in every utility function, but then I realized that this can be centralized inside the `WaitingFor` function.
I agree with you that usually a context should be provided, but I also think that a default one can be used. We are also providing a way to use an external one in case someone wants to use a custom one.